### PR TITLE
share/unshare commands

### PR DESCRIPTION
**Pull request by sigalmaya**
New commands to mange sharing instance: 
* share-instance [name] --id service-instance-id
* unshare-instance [name] --id service-instance-id
Sharing a service instance allows to consume it from various platforms.

### CODE REVIEW

File: internal/cmd/instance/update_sharing_instance.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		} else {
+			message = "Could not unshare service instance. Reason: "
+		}
+		output.PrintMessage(shc.Output, message)
+		return err
+	}
+
+	if len(location) != 0 {
+		if shc.share {
+			message = fmt.Sprintf("Service Instance %s successfully scheduled for sharing. To see status of the operation use:\n", shc.instanceName)
```
Comment (alexwo) [inline internal/cmd/instance/update_sharing_instance.go:113]: scheduled for sharing?

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+				})
+				Context("with id", func() {
+					It("should print object", func() {
+						validSyncUpdateSharingExect("myinstancename", "--id", "serviceinstanceid")
+						tableOutputExpected := instance.TableData().String()
+						Expect(buffer.String()).To(ContainSubstring(tableOutputExpected))
+					})
+				})
+
+				Context("With json output flag", func() {
```
Comment (alexwo) [inline internal/cmd/instance/update_sharing_instance_test.go:131]: with should be lower cased

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+					It("should be printed in json output format", func() {
+						validSyncUpdateSharingExect("instance-name", "--output", "json")
+						jsonByte, _ := json.MarshalIndent(instance, "", "  ")
+						jsonOutputExpected := string(jsonByte) + "\n"
+						Expect(buffer.String()).To(ContainSubstring(jsonOutputExpected))
+					})
+				})
+
+				Context("With yaml output flag", func() {
```
Comment (alexwo) [inline internal/cmd/instance/update_sharing_instance_test.go:140]: with should be lower cased

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+					})
+					It("should return error's description", func() {
+						err := invalidUpdateSharingCommandExecution("instance-name")
+						Expect(err).Should(HaveOccurred())
+						Expect(err.Error()).To(ContainSubstring("HTTP response error"))
+					})
+
+				})
+
+				Context("With invalid output format", func() {
```
Comment (alexwo) [inline internal/cmd/instance/update_sharing_instance_test.go:210]: with should be lower cased

File: internal/cmd/instance/update_sharing_instance.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+		result.Use = "share-instance [name] --id service-instance-id "
+		result.Short = "Share a service instance"
+		result.Long = `Share a service instance so that it can be consumed from various platforms in your subaccount.
+Instance can be shared only if it was created with the plan that supports instance sharing. For more information, see the documentation of the service whose instance you want to share.`
+	} else {
+		result.Use = "unshare-instance [name] --id service-instance-id "
+		result.Short = "Unshare a service instance"
+		result.Long = `Unshare a service instance to disable its consumption from any but the original platform in which it was created in your subaccount. If an instance you want to unshare has references, an error is returned`
+	}
+	result.Flags().StringVarP(&shc.instanceID, "id", "", "", "Id of the instance. Required in case when there are instances with same name")
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance.go:59]: The instance ID. Required if there is more than one instance with the same name.

File: internal/cmd/instance/update_sharing_instance.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+			FieldQuery: []string{
+				fmt.Sprintf("name eq '%s'", shc.instanceName),
+			},
+			GeneralParams: shc.Parameters.GeneralParams,
+		})
+		if err != nil {
+			return err
+		}
+		if len(instances.ServiceInstances) == 0 {
+			return fmt.Errorf("no instances found with name %s", shc.instanceName)
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance.go:88]: Couldn't find an instance with the name "%s".

File: internal/cmd/instance/update_sharing_instance.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		}
+
+		if len(instances.ServiceInstances) > 1 {
+			return fmt.Errorf("more than 1 instance found with name %s. Use --id flag to specify one", shc.instanceName)
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance.go:92]: Found more than one instance with the name "%s". Use the --id flag to specify which instance to update.

File: internal/cmd/instance/update_sharing_instance.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+		shc.instanceID = instances.ServiceInstances[0].ID
+	}
+	resultInstance, location, err := shc.Client.UpdateInstance(shc.instanceID, &types.ServiceInstance{
+		Shared: shc.share,
+	}, nil)
+	var message = ""
+	if err != nil {
+		if shc.share {
+			message = "Could not share service instance. Reason: "
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance.go:103]: Couldn't share the service instance. Reason:

File: internal/cmd/instance/update_sharing_instance.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+		} else {
+			message = "Could not unshare service instance. Reason: "
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance.go:105]: Couldn't unshare the service instance. Reason:

File: internal/cmd/instance/update_sharing_instance.go (hunk 1 above)
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance.go:113]: Service instance "%s" successfully scheduled for sharing. To see the status of the operation, use: \n", 

A question: is "%s" here a name or an ID?

File: internal/cmd/instance/update_sharing_instance.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+		output.PrintMessage(shc.Output, message)
+		return err
+	}
+
+	if len(location) != 0 {
+		if shc.share {
+			message = fmt.Sprintf("Service Instance %s successfully scheduled for sharing. To see status of the operation use:\n", shc.instanceName)
+
+		} else {
+			message = fmt.Sprintf("Service Instance %s successfully scheduled for unsharing. To see status of the operation use:\n", shc.instanceName)
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance.go:116]: Service instance "%s" successfully scheduled for unsharing. To see the status of the operation, use: \n, 

Same question as above for sharing.

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	}
+	type testCase struct {
+		share        bool
+		title        string
+		asyncMessage string
+		errorMessage string
+	}
+	tests := []testCase{
+		testCase{true, "share",
+			"successfully scheduled for sharing",
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance_test.go:80]: Successfully scheduled for sharing.

Do you need these?

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+			"Could not share service instance"},
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance_test.go:81]: Couldn't share the service instance.

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+		testCase{false, "unshare",
+			"successfully scheduled for unsharing",
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance_test.go:83]: Successfully scheduled for unsharing.

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+			"Could not unshare service instance"},
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance_test.go:84]: Couldn't unshare the service instance.

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+				Context("when service instance not found", func() {
+					BeforeEach(func() {
+						instances = &types.ServiceInstances{}
+					})
+
+					Context("by name", func() {
+						It("should return an error", func() {
+							err := invalidUpdateSharingCommandExecution("instance-name")
+							Expect(err.Error()).To(ContainSubstring(fmt.Sprintf("no instances found with name %s", "instance-name")))
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance_test.go:164]: Couldn't find an instance with the name "%s".

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+								},
+								types.ServiceInstance{
+									Name: "instance-name",
+								},
+							},
+						}, nil)
+					})
+					It("should return an error", func() {
+						err := invalidUpdateSharingCommandExecution("instance-name")
+						Expect(err.Error()).To(ContainSubstring(fmt.Sprintf("more than 1 instance found with name %s", "instance-name")))
```
Comment (i522006) [inline internal/cmd/instance/update_sharing_instance_test.go:185]: I already edited these above, right?

File: internal/cmd/instance/update_sharing_instance.go (hunk 1 above)
Comment (sigalmaya) [reply internal/cmd/instance/update_sharing_instance.go:113]: it's a name

Comment (i522006) [reply internal/cmd/instance/update_sharing_instance.go:113]: Ok, then, yes, we need the quotation marks.

File: internal/cmd/instance/update_sharing_instance_test.go (hunk 11 above)
Comment (sigalmaya) [reply internal/cmd/instance/update_sharing_instance_test.go:80]: no its a test

File: internal/cmd/instance/update_sharing_instance.go (hunk 10 above)
Comment (sigalmaya) [reply internal/cmd/instance/update_sharing_instance.go:116]: service name

Comment (i522006) [reply internal/cmd/instance/update_sharing_instance.go:116]: Ok, then we need " "

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+						err := invalidUpdateSharingCommandExecution("instance-name")
+						Expect(err).Should(HaveOccurred())
+						Expect(err.Error()).To(ContainSubstring("HTTP response error"))
+						Expect(buffer.String()).To(ContainSubstring(fmt.Sprintf("Couldn't %s the service instance. ", test.commandName)))
+
+					})
+
+				})
+
+				Context("with invalid output format", func() {
```
[the remark this answers is not in the document]
Comment (sigalmaya) [reply internal/cmd/instance/update_sharing_instance_test.go:185]: i added test for that and for -- mode  async

File: internal/cmd/instance/update_sharing_instance_test.go [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+				})
+			})
+
+			Context("invalid execution", func() {
+				JustBeforeEach(func() {
+					client.ListInstancesReturns(instances, nil)
+
+				})
+
+				Context("when service instance not found", func() {
```
[the remark this answers is not in the document]
Comment (sigalmaya) [reply internal/cmd/instance/update_sharing_instance_test.go:129]: fixed

**Review by EpsilonTal — APPROVED** [review]

**Pull request merged**